src/pip/_internal/network/auth.py

```diff
@@ -23,12 +25,52 @@
 
 logger = getLogger(__name__)
 
-Credentials = Tuple[str, str, str]
+
+class Credentials(NamedTuple):
+    service_name: str
+    username: str
+    password: str
+
+
+class KeyRingCli:
+    """Mirror the parts of keyring's API which pip uses
+
+    Instead of calling the keyring package installed alongside pip
+    we call keyring on the command line which will enable pip to
+    use which ever installation of keyring is available first in
+    PATH.
+    """
+
+    def __init__(self, keyring: str) -> None:
+        self.keyring = keyring
+
+    def get_password(self, service_name: str, username: str) -> Optional[str]:
+        cmd = [self.keyring, "get", service_name, username]
+        res = subprocess.run(
+            cmd,
+            stdin=subprocess.DEVNULL,
+            capture_output=True,
+            env=dict(PYTHONIOENCODING="utf-8"),
+        )
+        if res.returncode:
+            return None
+        return res.stdout.decode("utf-8").strip("\n")
+
+    def set_password(self, service_name: str, username: str, password: str) -> None:
+        cmd = [self.keyring, "set", service_name, username]
+        input_ = password.encode("utf-8") + b"\n"
+        res = subprocess.run(cmd, input=input_, env=dict(PYTHONIOENCODING="utf-8"))
+        res.check_returncode()
+        return None
+
 
 try:
     import keyring
 except ImportError:
     keyring = None  # type: ignore[assignment]
+    keyring_path = shutil.which("keyring")
+    if keyring_path is not None:
+        keyring = KeyRingCli(keyring_path)  # type: ignore[assignment]
```

Comment:
Instead of trying to fake keyring’s Python interface, it’d probably be easier if we introduce a common abstraction. We can have KeyRingCliProvider and KeyRingPythonProvider wrapping each implementation. This should also help get rid of the `ImportError` block.

Reply:
I've acted on this feedback